website/content/docs/enterprise/pkcs11-provider/index.mdx

```diff
@@ -0,0 +1,301 @@
+- `VAULT_LOG_FILE`: the location of the file the provider will use for logging. Defaults to standard out.
+- `VAULT_EMULATE_HARDWARE`: whether or not the provider will report that it is backed by a hardware device.
+
+## Encrypted TLS Key Support
+
+The TLS key returned by the KMIP engine is by default unencrypted.
+However, the PKCS#11 provider does support (limited) encryption options for the key using [RFC 1423](https://www.rfc-editor.org/rfc/rfc1423).
+We would only recommend using AES-256-CBC out of the available algorithms.
+
+The keys from KMIP should be ECDSA keys, and can be encrypted with a password with OpenSSL, e.g.,:
+
+```sh
+openssl ec -in cert.key -out encrypted.key -aes-256-cbc
+```
+
+The PKCS#11 provider will need access to the password to decrypt the TLS key.
+The password can be supplied to the provider in two ways:
+
+- The `VAULT_KMIP_KEY_PASSWORD` environment variable, or
+- the "PIN" parameter to the `C_Login` PKCS#11 function will be used to try to decrypt an encrypted TLS key.
+
+Note that only a single password can be supplied via the `VAULT_KMIP_KEY_PASSWORD`, so if multiple slots in the HCL file use encrypted TLS keys, they will either need to be encrypted with the same password, or use the `C_Login` method to specify the password.
+
+## Error Handling
+
+If an error occurs, the first place to check will be the `VAULT_LOG_FILE` for any relevant error messages.
+
+If the PKCS#11 provider returns an error code of `0x30` (`CKR_DEVICE_ERROR`), then an additional device error code may
+be available from the `C_SessionInfo` call.
+Here are the known device error codes the provider will return:
+
+| Code | Meaning                                                          |
+| ---- | ---------------------------------------------------------------- |
+| 400  | Invalid input was provided in the configuration or PKCS#11 call. |
+| 401  | Invalid credentials were provided.                               |
+| 404  | The object, attribute, or key was not found.                     |
+| 600  | An unknown I/O error occurred.                                   |
+| 601  | A KMIP engine error occured.                                     |
+
+## Capabilities
+
+The Vault PKCS#11 provider implements the following PKCS#11 provider profiles:
+
+- [Baseline](http://docs.oasis-open.org/pkcs11/pkcs11-profiles/v2.40/os/pkcs11-profiles-v2.40-os.html#_Toc416960548)
+- [Extended](http://docs.oasis-open.org/pkcs11/pkcs11-profiles/v2.40/os/pkcs11-profiles-v2.40-os.html#_Toc416960554)
+
+The following mechanisms are currently supported, as of Vault 1.12:
+
+| Name               | Mechanism Number |
+| ------------------ | ---------------- |
+| AES key generation | `0x1080`         |
+| AES-CBC            | `0x1082`         |
+| AES-CBC Pad        | `0x1085`         |
+| AES-CTR            | `0x1086`         |
+| AES-GCM            | `0x1087`         |
```

Comment:
I'm curious where these mechanism numbers come from? I didn't seem them in the Baseline and Extended links above.

Reply:
You want the `pkcs11t.h` file of some distributions: https://github.com/nss-dev/nss/blob/master/lib/util/pkcs11t.h 

http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/errata01/os/include/pkcs11-v2.40/pkcs11t.h

The values in the second header file are definitive; the HTML/... spec sometimes has wrong values and they always defer to the value in the shipped header file.

Reply:
The `pkcs11t.h` file is also in this repository, as that is where we get the numbers from in the library.